packages/require/lib/compiler.ts

```diff
@@ -0,0 +1,94 @@
+import path from "path";
+import TruffleError from "@truffle/error";
+
+import type { CreateOptions, Service } from "ts-node";
+import TruffleConfig from "@truffle/config";
+import { readFileSync, existsSync } from "fs";
+
+import originalRequire from "original-require";
+
+import vm from "vm";
+
+/**
+ * @returns compiler function from user's installation of `ts-node` if script is
+ * a ts or tsx file, otherwise `undefined`.
+ */
+export function compile(conf: TruffleConfig, scriptPath: string): string {
+  const source = readFileSync(scriptPath, { encoding: "utf-8" });
+
+  // we only compile TS files, so just return the source unless we are dealing
+  // with one of those
+  if (![".ts", ".tsx", ".cts"].includes(path.extname(scriptPath))) {
+    return source;
+  }
+
+  const compilationService = _getOrCreateCompilationService(conf, scriptPath);
+
+  return compilationService.compile(source, scriptPath);
+}
+
+export function registerTsNode(
+  conf: TruffleConfig,
+  scriptPath: string,
+  context: Object
+) {
+  // we only compile TS files, so bail unless we are dealing with one of those;
+  if (![".ts", ".tsx", ".cts"].includes(path.extname(scriptPath))) {
+    return;
+  }
+
+  const code = `
+    const tsNode = require("ts-node");
+    const registrationOptions = {
+      cwd: ${JSON.stringify(conf.working_directory.toString())},
+      compilerOptions: {
+        inlineSourceMap: true
+      }
+    };
+    tsNode.register(registrationOptions);
+    `;
+  const script = new vm.Script(code);
+  script.runInContext(context);
+}
+
+let _compilationService: Service | null = null;
+
+function _getOrCreateCompilationService(
+  conf: TruffleConfig,
+  scriptPath: string
+): Service {
+  if (!_compilationService) {
+    try {
+      const createOptions: CreateOptions = {
+        cwd: path.dirname(scriptPath),
+        esm: false,
+        compilerOptions: {
+          inlineSourceMap: true
+        }
+      };
+
+      _compilationService = originalRequire("ts-node").create(createOptions);
+
+      const globalDeclarationsPath = existsSync(
+        path.join(__dirname, "scriptGlobals.ts")
+      )
+        ? path.join(__dirname, "scriptGlobals.ts")
+        : path.join(__dirname, "..", "scriptGlobals.ts");
+      _compilationService!.compile(
+        readFileSync(globalDeclarationsPath, { encoding: "utf-8" }),
+        globalDeclarationsPath
+      );
```

Comment:
This adds the typings for the global context to the cached `compilationService` instance. I tested doing this a few different ways, and this method gave me the fewest (possibly even zero?) compromises as compared to the others.

Doing it this way and caching the service lets us read in the typings only once (rather than for each TS file we that we interpret), and it doesn't prepend data to the user's source, which avoids needing to deal with line number offsets when reporting type check, compilation, and runtime errors.

Reply:
Also note that for now these typings are static and don't change depending on what context the caller passes in. For now I'd like to take that as tech debt, as solving that issue wasn't necessary to make any of the existing code paths work.

**Please advise if you disagree and think that it should be included as part of this change.** I just didn't want to bite it off now, as while I have what looks to be a straightforward plan-of-attack for it, I encountered a bunch of unexpected problems while making just this change work, and I didn't want to delay getting TS migrations out to users any longer than absolutely necessary.